overlord/snapstate/backend/link.go

```diff
@@ -0,0 +1,152 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2014-2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package backend
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/ubuntu-core/snappy/logger"
+	"github.com/ubuntu-core/snappy/progress"
+	"github.com/ubuntu-core/snappy/snap"
+	// XXX: eventually not needed
+	"github.com/ubuntu-core/snappy/snappy"
+	"github.com/ubuntu-core/snappy/wrappers"
+)
+
+func updateCurrentSymlinks(info *snap.Info) error {
+	mountDir := info.MountDir()
+
+	currentActiveSymlink := filepath.Join(mountDir, "..", "current")
+	if err := os.Remove(currentActiveSymlink); err != nil && !os.IsNotExist(err) {
+		logger.Noticef("Cannot remove %q: %v", currentActiveSymlink, err)
+	}
+
+	dataDir := info.DataDir()
+	currentDataSymlink := filepath.Join(dataDir, "..", "current")
+	if err := os.Remove(currentDataSymlink); err != nil && !os.IsNotExist(err) {
+		logger.Noticef("Cannot remove %q: %v", currentDataSymlink, err)
+	}
+
+	if err := os.MkdirAll(info.DataDir(), 0755); err != nil {
+		return err
+	}
+
+	if err := os.Symlink(filepath.Base(dataDir), currentDataSymlink); err != nil {
+		return err
+	}
+
+	return os.Symlink(filepath.Base(mountDir), currentActiveSymlink)
+}
+
+// LinkSnap makes the snap available by generating wrappers and setting the current symlinks.
+func (b Backend) LinkSnap(info *snap.Info) error {
+	if err := generateWrappers(info); err != nil {
+		return err
+	}
+
+	// XXX/TODO: this needs to be a task with proper undo and tests!
+	if err := snappy.SetNextBoot(info); err != nil {
+		return err
+	}
+
+	return updateCurrentSymlinks(info)
+}
+
+func generateWrappers(s *snap.Info) error {
```

Comment:
@zyga as I mention in the description of the PR (#1134) this is from, we should look into logging more into tasks which probably would require more symmetry, atm the asymmetry is because only when stopping services the old code was worried about notifiying the user of the wait